doing.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "DOING" "1" "September 2021" "" ""
.
.SH "SYNOPSIS"
\fBdoing [global options] command [command options] [arguments\.\.\.]\fR
.
.SH "DESCRIPTION"
\fIA command line tool for remembering what you were doing and tracking what you\'ve done\.\fR
.
.P
\fBdoing\fR is a basic CLI for adding and listing "what was I doing" reminders in a TaskPaper\-formatted \fIhttps://www\.taskpaper\.com\fR text file\. It allows for multiple sections/categories and flexible output formatting\.
.
.SH "USAGE"
.
.nf

doing [global options] command [command options] [arguments\.\.\.]
.
.fi
.
.SS "GLOBAL OPTIONS:"
.
.nf

\-f, \-\-doing_file=arg \- Specify a different doing_file (default: none)
\-\-help               \- Show this message
\-\-[no\-]notes         \- Output notes if included in the template (default: enabled)
\-\-stdout             \- Send results report to STDOUT instead of STDERR
\-\-version            \- Display the program version
\-x, \-\-[no\-]noauto    \- Exclude auto tags and default tags
.
.fi
.
.SS "COMMANDS:"
.
.nf

help           \- Shows a list of commands and global options
help [command] \- Shows help for any command (`doing help now`)


add_section   \- Add a new section to the "doing" file
again, resume \- Repeat last entry as new entry
archive       \- Move entries between sections
cancel        \- End last X entries with no time tracked
choose        \- Select a section to display from a menu
colors        \- List available color variables for configuration templates and views
config        \- Edit the configuration file
done, did     \- Add a completed item with @done(date)\. No argument finishes last entry\.
finish        \- Mark last X entries as @done
grep, search  \- Search for entries
import        \- Import entries from an external source
last          \- Show the last entry, optionally edit
later         \- Add an item to the Later section
mark, flag    \- Mark last entry as highlighted
meanwhile     \- Finish any running @meanwhile tasks and optionally create a new one
note          \- Add a note to the last entry
now, next     \- Add an entry
on            \- List entries for a date
open          \- Open the "doing" file in an editor
recent        \- List recent entries
sections      \- List sections
select        \- Display an interactive menu to perform operations (requires fzf)
show          \- List all entries
since         \- List entries since a date
tag           \- Add tag(s) to last entry
template      \- Output HTML and CSS templates for customization
today         \- List entries from today
undo          \- Undo the last change to the doing_file
view          \- Display a user\-created view
views         \- List available custom views
yesterday     \- List entries from yesterday
.
.fi
.
.SH "CONFIGURATION"
Run \fBdoing config\fR to open your \fB~/\.doingrc\fR file in the editor defined in the \fB$EDITOR\fR environment variable\. Set up your \fBdoing_file\fR right away (where you want entries to be stored), and cover the rest after you\'ve read the docs\.
.
.P
See the \fISUPPORT\fR section below for troubleshooting details\.
.
.SS "THE DOING FILE"
The file that stores all of your entries is generated the first time you add an entry (with \fBdoing now\fR or \fBdoing later\fR)\. By default, the file is created in \fB~/what_was_i_doing\.md\fR, but you can modify this in the config file\.
.
.P
The format of the "doing" file is TaskPaper\-compatible\. You can edit it by hand at any time (in TaskPaper or any text editor), but it uses a specific format for parsing, so be sure to maintain the dates and pipe characters\.
.
.P
Notes are anything in the list without a leading hyphen and date\. They belong to the entry directly before them, and they should be indented one level beyond the parent item\.
.
.P
When using the \fBnow\fR and \fBlater\fR commands on the command line, you can start the entry with a quote and hit return, then type the note and close the quote\. Anything after the first line will be turned into a TaskPaper\-compatible note for the task and can be displayed in templates using \fB%note\fR\.
.
.P
Notes can be prevented from ever appearing in output with the global option \fB\-\-no\-notes\fR: \fBdoing \-\-no\-notes show all\fR\.
.
.P
Auto tagging (adding tags listed in \.doingrc under \fBautotag\fR and \fBdefault_tags\fR) can be skipped for an entry with the \fB\-x\fR global option: \fBdoing \-x done skipping some automatic tagging\fR\.
.
.SS "CONFIGURATION OPTIONS"
A basic configuration looks like this:
.
.IP "" 4
.
.nf

\-\-\-
doing_file: /Users/username/Dropbox/doing\.taskpaper
current_section: Currently
default_template: \'%date: %title%note\'
default_date_format: \'%Y\-%m\-%d %H:%M\'
marker_tag: flagged
marker_color: yellow
tags_color: boldcyan
default_tags: []
editor_app: TaskPaper
config_editor_app: Sublime Text
:include_notes: true
views:
  color:
    date_format: \'%F %_I:%M%P\'
    section: Currently
    count: 10
    wrap_width: 0
    template: \'%boldblack%date %boldgreen| %boldwhite%title%default%note\'
    order: desc
templates:
  default:
    date_format: \'%Y\-%m\-%d %H:%M\'
    template: \'%date | %title%note\'
    wrap_width: 0
  today:
    date_format: \'%_I:%M%P\'
    template: \'%date: %title%odnote\'
    wrap_width: 0
  last:
    date_format: \'%_I:%M%P on %a\'
    template: \'%title (at %date)%odnote\'
    wrap_width: 0
  recent:
    date_format: \'%_I:%M%P\'
    template: \'%date > %title%odnote\'
    wrap_width: 50
autotag:
  whitelist:
  \- coding
  \- design
  synonyms:
    brainstorming:
    \- thinking
    \- idea
html_template:
  haml:
  css:
.
.fi
.
.IP "" 0
.
.P
The config file is stored in \fB~/\.doingrc\fR, and a skeleton file is created on the first run\. Just run \fBdoing\fR on its own to create the file\.
.
.SS "PER\-FOLDER CONFIGURATION"
Any options found in a \fB\.doingrc\fR anywhere in the hierarchy between your current folder and your home folder will be appended to the base configuration, overriding or extending existing options\. This allows you to put a \fB\.doingrc\fR file into the base of a project and add specific configurations (such as default tags) when working in that project on the command line\. These can be cascaded, with the closest \fB\.doingrc\fR to your current directory taking precedence, though I\'m not sure why you\'d want to deal with that\.
.
.P
Possible uses:
.
.IP "\(bu" 4
Define custom HTML output on a per\-project basis using the html_template option for custom templates\. Customize time tracking reports based on project or client\.
.
.IP "\(bu" 4
Define \fBdefault_tags\fR for a project so that every time you \fBdoing now\fR from within that project directory or its subfolders, it gets tagged with that project automatically\.
.
.IP "" 0
.
.P
Any part of the configuration can be copied into these local files and modified\. You only need to include the parts you want to change or add\.
.
.SS "DOING FILE LOCATION"
The one thing you\'ll probably want to adjust is the file that the notes are stored in\. That\'s the \fBdoing_file\fR key:
.
.IP "" 4
.
.nf

doing_file: /Users/username/Dropbox/nvALT2\.2/?? What was I doing\.md
.
.fi
.
.IP "" 0
.
.P
I keep mine in my nvALT folder for quick access and syncing between machines\. If desired, you can give it a \fB\.taskpaper\fR extension to make it more recognizable to other applications\. (If you do that in nvALT, make sure to add \fBtaskpaper\fR as a recognized extension in preferences)\.
.
.SS "CURRENT ACTIONS SECTION"
You can rename the section that holds your current tasks\. By default, this is \fBCurrently\fR, but if you have some other bright idea, feel free:
.
.IP "" 4
.
.nf

current_section: Currently
.
.fi
.
.IP "" 0
.
.SS "DEFAULT EDITORS"
The setting \fBeditor_app\fR only applies to Mac OS X users\. It\'s the default application that the command \fBdoing open\fR will open your WWID file in\. Personally, I use \fBeditor_app: TaskPaper\fR\. If this is blank, it will be opened by whatever the system default is for the doing file extension (default is \fB\.md\fR), or you can use \fB\-a app_name\fR or \fB\-b bundle_id\fR to override\.
.
.P
You can define a separate app to open the config file in when running \fBdoing config\fR\. The key for this is \fBconfig_editor_app\fR\.
.
.P
In the case of the \fBdoing now \-e\fR command, your \fB$EDITOR\fR environment variable will be used to complete the entry text and notes\. Set it in your \fB~/\.bash_profile\fR or whatever is appropriate for your system:
.
.IP "" 4
.
.nf

export EDITOR="mate \-w"
.
.fi
.
.IP "" 0
.
.P
The only requirements are that your editor be launchable from the command line and able to "wait\." In the case of Sublime Text and TextMate, just use \fB\-w\fR like this: \fBexport EDITOR="subl \-w"\fR\.
.
.SS "TEMPLATES"
The config also contains templates for various command outputs\. Include placeholders by placing a % before the keyword\. The available tokens are:
.
.IP "\(bu" 4
\fB%title\fR: the "what was I doing" entry line
.
.IP "\(bu" 4
\fB%date\fR: the date based on the template\'s \fBdate_format\fR setting
.
.IP "\(bu" 4
\fB%shortdate\fR: a custom date formatter that removes the day/month/year from the entry if they match the current day/month/year
.
.IP "\(bu" 4
\fB%note\fR: Any note in the entry will be included here, a newline and tabs are automatically added\.
.
.IP "\(bu" 4
\fB%odnote\fR: The notes with a leading tab removed (outdented note)
.
.IP "\(bu" 4
\fB%chompnote\fR: Notes on one line, beginning and trailing whitespace removed\.
.
.IP "\(bu" 4
\fB%section\fR: The section/project the entry is currently in
.
.IP "\(bu" 4
\fB%hr\fR: a horizontal rule (\fB\-\fR) the width of the terminal
.
.IP "\(bu" 4
\fB%hr_under\fR: a horizontal rule (\fB_\fR) the width of the terminal
.
.IP "\(bu" 4
\fB%n\fR: inserts a newline
.
.IP "\(bu" 4
\fB%t\fR: inserts a tab
.
.IP "\(bu" 4
.
.IP "\(bu" 4
you can prefix \fBbg\fR to affect background colors (\fB%bgyellow\fR)
.
.IP "\(bu" 4
prefix \fBbold\fR and \fBboldbg\fR for strong colors (\fB%boldgreen\fR, \fB%boldbgblue\fR)
.
.IP "\(bu" 4
there are some random special combo colors\. Use \fBdoing colors\fR to see the list
.
.IP "" 0

.
.IP "\(bu" 4
\fB%interval\fR: when used with the \fB\-t\fR switch on the \fBshow\fR command, it will display the time between a timestamp or \fI@start(date)\fR tag and the \fI@done(date)\fR tag, if it exists\. Otherwise, it will remain empty\.
.
.IP "" 0
.
.P
Date formats are based on Ruby \fBstrftime\fR \fIhttp://www\.ruby\-doc\.org/stdlib\-2\.1\.1/libdoc/date/rdoc/Date\.html#method\-i\-strftime\fR formatting\.
.
.P
My normal template for the \fBrecent\fR command looks like this:
.
.IP "" 4
.
.nf

recent:
  date_format: \'%_I:%M%P\'
  template: \'%date > %title%odnote\'
  wrap_width: 88
.
.fi
.
.IP "" 0
.
.P
And it outputs:
.
.IP "" 4
.
.nf

$ doing recent 3
 4:30am > Made my `console` script smarter\.\.\.
    Checks first argument to see if it\'s a file, if it is, that\'s the log

    Else, it checks the first argument for a "\.log" suffix and does a search in the user
    application logs with `find` for it\.

    Otherwise, system\.log\.

    I also made an awesome Cope wrapper for it\.\.\.
12:00pm > Working on `doing` again\.
12:45pm > I think this thing (doing) is ready to document and distribute
$
.
.fi
.
.IP "" 0
.
.P
You can get pretty clever and include line breaks and other formatting inside of double quotes\. If you want multiline templates, just use \fB\en\fR in the template line, and after the next run it will be rewritten as proper YAML automatically\.
.
.P
For example, this block:
.
.IP "" 4
.
.nf

recent:
  date_format: \'%_I:%M%P\'
  template: "\en%hr\en%date\en > %title%odnote\en%hr_under"
  wrap_width: 100
.
.fi
.
.IP "" 0
.
.P
will rewrite to:
.
.IP "" 4
.
.nf

recent:
  date_format: \'%_I:%M%P\'
  template: |2\-

    %hr
    %date
     > %title%odnote
    %hr_under
  wrap_width: 100
.
.fi
.
.IP "" 0
.
.P
and output my recent entries like this:
.
.IP "" 4
.
.nf

$ doing recent 3
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
 4:30am
 > Made my `console` script smarter\.\.\.
    Checks first argument to see if it\'s a file, if it is, that\'s the log

    Else, it checks the first argument for a "\.log" suffix and does a search in the user application
    logs with `find` for it\.

    Otherwise, system\.log\.

    I also made an awesome Cope wrapper for it\.\.\.
_______________________________________________________________________

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
12:00pm
 > Working on `doing` again\.
_______________________________________________________________________

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
12:45pm
 > I think this thing (doing) is ready to document and distribute
_______________________________________________________________________

$
.
.fi
.
.IP "" 0
.
.SS "CUSTOM VIEWS"
You can create your own "views" in the \fB~/\.doingrc\fR file and view them with \fBdoing view view_name\fR\. Just add a section like this:
.
.IP "" 4
.
.nf

views:
  old:
    section: Old
    count: 5
    wrap_width: 0
    date_format: \'%F %_I:%M%P\'
    template: \'%date | %title%note\'
    order: asc
    tags: done finished cancelled
    tags_bool: ANY
.
.fi
.
.IP "" 0
.
.P
You can add additional custom views\. Just nest them under the \fBviews\fR key (indented two spaces from the edge)\. Multiple views would look like this:
.
.IP "" 4
.
.nf

views:
  later:
    section: Later
    count: 5
    wrap_width: 60
    date_format: \'%F %_I:%M%P\'
    template: \'%date | %title%note\'
  old:
    section: Old
    count: 5
    wrap_width: 0
    date_format: \'%F %_I:%M%P\'
    template: \'%date | %title%note\'
.
.fi
.
.IP "" 0
.
.P
The \fBsection\fR key is the default section to pull entries from\. Count and section can be overridden at runtime with the \fB\-c\fR and \fB\-s\fR flags\. Setting \fBsection\fR to \fBAll\fR will combine all sections in the output\.
.
.P
You can add new sections with \fBdoing add_section section_name\fR\. You can also create them on the fly by using the \fB\-s section_name\fR flag when running \fBdoing now\fR\. For example, \fBdoing now \-s Misc just a random side note\fR would create the "just a random side note" entry in a new section called "Misc," if Misc didn\'t already exist\.
.
.P
The \fBtags\fR and \fBtags_bool\fR keys allow you to specify tags that the view is filtered by\. You can list multiple tags separated by spaces, and then use \fBtags_bool\fR to specify \fBALL\fR, \fBANY\fR, or \fBNONE\fR to determine how it handles the multiple tags\.
.
.P
The \fBorder\fR key defines the sort order of the output\. This is applied \fIafter\fR the tasks are retrieved and cut off at the maximum number specified in \fBcount\fR\.
.
.P
Regarding colors, you can use them to create very nice displays if you\'re outputting to a color terminal\. Example:
.
.IP "" 4
.
.nf

color:
  date_format: \'%F %_I:%M%P\'
  section: Currently
  count: 10
  wrap_width: 0
  template: \'%boldblack%date %boldgreen| %boldwhite%title%default%note\'
.
.fi
.
.IP "" 0
.
.P
You can also specify a default output format for a view\. Most of the optional output formats override the template specification (\fBhtml\fR, \fBcsv\fR, \fBjson\fR)\. If the \fBview\fR command is used with the \fB\-o\fR flag, it will override what\'s specified in the file\.
.
.SS "COLORS"
You can use the following colors in view templates\. Set a foreground color with a named color:
.
.IP "" 4
.
.nf

%black
%red
%green
%yellow
%blue
%magenta
%cyan
%white
.
.fi
.
.IP "" 0
.
.P
You can also add a background color (\fB%bg[color]\fR) by placing one after the foreground color:
.
.IP "" 4
.
.nf

%white%bgblack
%black%bgred
\.\.\.etc\.
.
.fi
.
.IP "" 0
.
.P
There are bold variants for both foreground and background colors
.
.IP "" 4
.
.nf

%boldblack
%boldred
\.\.\. etc\.

%boldbgblack
%boldbgred
\.\.\. etc\.
.
.fi
.
.IP "" 0
.
.P
And a few special colors you\'ll just have to try out to see (or just run \fBdoing colors\fR):
.
.IP "" 4
.
.nf

%softpurple
%hotpants
%knightrider
%flamingo
%yeller
%whiteboard
.
.fi
.
.IP "" 0
.
.P
Any time you use one of the foreground colors it will reset the bold and background settings to their default automatically\. You can force a reset to default terminal colors using \fB%default\fR\.
.
.SS "HTML TEMPLATES"
For commands that provide an HTML output option, you can customize the templates used for markup and CSS\. The markup uses HAML \fIhttp://haml\.info/\fR, and the styles are pure CSS\.
.
.P
To export the default configurations for customization, use \fBdoing templates \-\-type=[HAML|CSS]\fR\. This will output to STDOUT where you can pipe it to a file, e\.g\. \fBdoing templates \-\-type=HAML > my_template\.haml\fR\. You can modify the markup, the CSS, or both\.
.
.P
Once you have either or both of the template files, edit \fB\.doingrc\fR and look for the \fBhtml_template:\fR section\. There are two subvalues, \fBhaml:\fR and \fBcss:\fR\. Add the path to the templates you want to use\. A tilde may be substituted for your home directory, e\.g\. \fBcss: ~/styles/doing\.css\fR\.
.
.SS "AUTOTAGGING"
Keywords in your entries can trigger automatic tagging, just to make life easier\. There are three tools available: default tags, whitelisting, and synonym tagging\.
.
.P
Default tags are tags that are applied to every entry\. You probably don\'t want to add these in the root configuration, but using a local \fB\.doingrc\fR in a project directory that defines default tags for that project allows anything added from that directory to be tagged automatically\. A local \fB\.doingrc\fR in my Marked development directory might contain:
.
.IP "" 4
.
.nf

\-\-\-
default_tags: [marked,coding]
.
.fi
.
.IP "" 0
.
.P
And anything I enter while in the directory gets tagged with \fI@marked\fR and \fI@coding\fR\.
.
.P
A whitelist is a list of words that should be converted directly into \fI@tags\fR\. If my whitelist contains "design" and I type \fBdoing now working on site design\fR, that\'s automatically converted to "working on site @design\."
.
.P
Synonyms allow you to define keywords that will trigger their parent tag\. If I have a tag called \fI@design\fR, I can add "typography" as a synonym\. Then entering \fBdoing now working on site typography\fR will become "working on site typography @design\."
.
.P
White lists and synonyms are defined like this:
.
.IP "" 4
.
.nf

autotag:
  synonyms:
    design:
    \- typography
    \- layout
    brainstorming
    \- thinking
    \- idea
  whitelist:
  \- brainstorming
  \- coding
.
.fi
.
.IP "" 0
.
.P
Note that you can include a tag with synonyms in the whitelist as well to tag it directly when used\.
.
.P
\fI\fBAdding entries:\fR\fR
.
.IP "" 4
.
.nf

now, did      \- Add an entry
later         \- Add an item to the Later section
done          \- Add a completed item with @done(date)\. No argument finishes last entry\.
meanwhile     \- Finish any @meanwhile tasks and optionally create a new one
again, resume \- Duplicate the last entry as new entry (without @done tag)
.
.fi
.
.IP "" 0
.
.P
The \fBdoing now\fR command can accept \fB\-s section_name\fR to send the new entry straight to a non\-default section\. It also accepts \fB\-\-back=AMOUNT\fR to let you specify a start date in the past using "natural language\." For example, \fBdoing now \-\-back=25m ENTRY\fR or \fBdoing now \-\-back="yesterday 3:30pm" ENTRY\fR\.
.
.P
If you want to use \fB\-\-back\fR with \fBdoing done\fR but want the end time to be different than the start time, you can either use \fB\-\-took\fR in addition, or just use \fB\-\-took\fR on its own as it will backdate the start time such that the end time is now and the duration is equal to the value of the \fB\-\-took\fR argument\.
.
.P
You can finish the last unfinished task when starting a new one using \fBdoing now\fR with the \fB\-f\fR switch\. It will look for the last task not marked \fI@done\fR and add the \fI@done\fR tag with the start time of the new task (either the current time or what you specified with \fB\-\-back\fR)\.
.
.P
\fBdoing done\fR is used to add an entry that you\'ve already completed\. Like \fBnow\fR, you can specify a section with \fB\-s section_name\fR\. You can also skip straight to Archive with \fB\-a\fR\.
.
.P
\fBdoing done\fR can also backdate entries using natural language with \fB\-\-back 15m\fR or \fB\-\-back "3/15 3pm"\fR\. That will modify the starting timestamp of the entry\. You can also use \fB\-\-took 1h20m\fR or \fB\-\-took 1:20\fR to set the finish date based on a "natural language" time interval\. If \fB\-\-took\fR is used without \fB\-\-back\fR, then the start date is adjusted (\fB\-\-took\fR interval is subtracted) so that the completion date is the current time\.
.
.P
When used with \fBdoing done\fR, \fB\-\-back\fR and \fB\-\-took\fR allow time intervals to be accurately counted when entering items after the fact\. \fB\-\-took\fR is also available for the \fBdoing finish\fR command, but cannot be used in conjunction with \fB\-\-back\fR\. (In \fBfinish\fR they both set the end date, and neither has priority\. \fB\-\-back\fR allows specific days/times, \fB\-\-took\fR uses time intervals\.)
.
.P
All of these commands accept a \fB\-e\fR argument\. This opens your command line editor (as defined in the environment variable \fB$EDITOR\fR)\. Add your entry, save the temp file, and close it\. The new entry is added\. Anything after the first line is included as a note on the entry\.
.
.P
\fBdoing again\fR (or \fBdoing resume\fR) will duplicate the last @done entry (most recently completed) with a new start date (and without the @done tag)\. To resume the last entry matching specific tags, use \fB\-\-tag=TAG\fR\. You can specify multiple tags by separating with a comma\. Multiple tags are combined with \'AND\' by default (all tags must exist on the entry to match), but you can use \fB\-\-bool=\fR to set it to \'OR\' or \'NOT\'\. By default the new entry will be added to the same section as the matching entry, but you can specify a section with \fB\-\-in=SECTION\fR\.
.
.P
\fBdoing meanwhile\fR is a special command for creating and finishing tasks that may have other entries come before they\'re complete\. When you create an entry with \fBdoing meanwhile [entry text]\fR, it will automatically complete the last \fI@meanwhile\fR item (dated \fI@done\fR tag) and add the \fI@meanwhile\fR tag to the new item\. This allows time tracking on a more general basis, and still lets you keep track of the smaller things you do while working on an overarching project\. The \fBmeanwhile\fR command accepts \fB\-\-back [time]\fR and will backdate the \fI@done\fR tag and start date of the new task at the same time\. Running \fBmeanwhile\fR with no arguments will simply complete the last \fI@meanwhile\fR task\.
.
.P
See \fBdoing help meanwhile\fR for more options\.
.
.P
\fI\fBModifying entries:\fR\fR
.
.IP "" 4
.
.nf

finish      \- Mark last X entries as @done
cancel      \- Mark last X entries as @done without completion date
tag         \- Tag last entry
note        \- Add a note to the last entry
.
.fi
.
.IP "" 0
.
.P
\fI\fBFinishing\fR\fR
.
.P
\fBdoing finish\fR by itself is the same as \fBdoing done\fR by itself\. It adds \fI@done(timestamp)\fR to the last entry\. It also accepts a numeric argument to complete X number of tasks back in history\. Add \fB\-a\fR to also archive the affected entries\.
.
.P
\fBdoing finish\fR also provides an \fB\-\-auto\fR flag, which you can use to set the end time of any entry to 1 minute before the start time of the next\. Running a command such as \fBdoing finish \-\-auto 10\fR will go through the last 10 entries and sequentially update any without a \fI@done\fR tag with one set to the time just before the next entry in the list\.
.
.P
As mentioned above, \fBfinish\fR also accepts \fB\-\-back "2 hours"\fR (sets the finish date from time now minus interval) or \fB\-\-took 30m\fR (sets the finish date to time started plus interval) so you can accurately add times to completed tasks, even if you don\'t do it in the moment\.
.
.P
Both \fBdone\fR and \fBfinish\fR accept an \fB\-\-archive\fR switch which immediately moves the completed entries to the Archive section with a \fB@from(Project)\fR tag\.
.
.P
You can finish the last entry containing a specific tag or combination of tags using the \fB\-\-tag\fR flag\. Multiple tags are separated by commas\. By default tags are combined with an AND boolean, meaning the entry must contain all specified tags to be considered\. For example, to finish the last entry containing both "@work" and "@project1", you would use:
.
.IP "" 4
.
.nf

doing finish \-\-tag=work,project1
.
.fi
.
.IP "" 0
.
.P
You can change the boolean using \fB\-\-bool=OR\fR (last entry containing any of the specified tags) or \fB\-\-bool=NOT\fR (last entry containing none of the tags)\.
.
.P
You can also include a \fB\-\-no\-date\fR switch to add \fB@done\fR without a finish date, meaning no time is tracked for the task\. \fBdoing cancel\fR is an alias for this\. Like \fBfinish\fR, \fBcancel\fR accepts a count to act on the last X entries, as well as \fB\-\-archive\fR and \fB\-\-section\fR options\. \fBcancel\fR also accepts the \fB\-\-tag\fR and \fB\-\-bool\fR flags for tag filtering\.
.
.P
\fI\fBTagging and Autotagging\fR\fR
.
.P

CHANGELOG.md
### 1.0.82

- Bugfixes

### 1.0.81

- fzf menu improvements
- allow multiple selections `doing select` action menu

### 1.0.80

- Convert all menus to fzf screens

### 1.0.79

- Gem missing fzf
- Wildcard tag removal using `doing select -t "tag*" -r`
- fzf menu display polish

### 1.0.78

- If no action is specified with select command, an interactive menu is
presented
- add output action select command with formatting and save options
- Don't link urls in html output that don't have a protocol

### 1.0.76

- Refine editing multiple selections (doing select)

### 1.0.74

- Add --tag and --search flags to tag command to tag all entries matching search terms
- Add since command, which is the same as `doing on tuesday to now` but `doing since tuesday` just feels more intuitive. 

### 1.0.73

- Fix for timeline output

### 1.0.72

- Add `doing select` to show menu of all tasks, searchable with fuzzy matching and the ability to perform certain tasks on multiple selections.

### 1.0.71

- Fix for template command not working at all

### 1.0.70

- Fix for `doing done --took 30m` setting the wrong @done timestamp when completing previous item

### 1.0.69

- Add `--unfinished` option to finish and cancel commands

### 1.0.68

- Fix error in `doing show --sort` argument parsing

### 1.0.67

- Gem packaging error

### 1.0.66

- Fix for some long flags being interpreted as arrays instead of strings
- More flexible boolean specification, can be: all, and, any, or, not, or none
- Fix for archive command not removing original entries from archived section

### 1.0.65

- Prevent duplicates/overlapping entries when importing

### 1.0.64

- Initial import feature for Timing.app reports

### 1.0.63

- README updates
- If `doing done --took=X` results in completion date greater than current time, use current time as completion date and backdate the entry's timestamp to fit

### 1.0.62

- Fix: `doing done` with `--took=` and without `--back=` should set end time to start date plus `--took` value

### 1.0.61

- Add --search filter to `doing archive`

### 1.0.60

#### FIXED

- Default value for `doing again --bool` was ALL, should be AND

### 1.0.59

#### IMPROVED

- Improvements to `doing again --tag=` functionality

### 1.0.58

#### IMPROVED

- Finish previous task if `doing again` and not already completed

### 1.0.57

#### IMPROVED

- Unit tests

### 1.0.56

#### IMPROVED

- Tag command tests

#### FIXED

- Doing not reading per-directory .doingrc configs

### 1.0.55

#### NEW

- Added config_editor_app setting to config so you can have
- A parenthetical at the end of an entry title becomes an attached
- `--editor` flag for `doing last` to edit last entry
- `--tag=` flag to filter `doing last` by tag
- `--search=` to filter `doing last` by text/regex search
- `--search=` for `doing finish`, finish last X entries matching search
- Add `tags_color` as a primary config key to highlight @tags in displayed entries

#### IMPROVED

- Clean up command line help
- --editor improvements for all commands that use it

#### FIXED

- Doing finish --took throwing error
- Doing tag --remove was adding tags if they didn't exist
- Creating a meanwhile task with a note resulted in an error

### 1.0.54

#### FIXED

- Bugfix for `finish --tag=TAG`

### 1.0.53

#### NEW

- `--tag` and `--bool` filtering for again/resume, cancel
- `--in` flag for `again`/`resume` to specify to which section the new
- Finish command accepts `--tag=` flag, finishing last entry
- `doing cancel` to end X tasks without completion date (alias for

#### IMPROVED

- Add --no-color option to view command
- Add --tag to show for compatibility

#### FIXED

- Error running finish without --tag flag
- --archive flag on finish, done, and cancel causing error

### 1.0.52

#### NEW

- Finish command accepts `--tag=` flag, finishing last entry
- `doing cancel` to end X tasks without completion date (alias for

#### FIXED

- --archive flag on finish, done, and cancel causing error

### 1.0.49

- Fix for missing date on @done tags

### 1.0.48

- Fix confirmation dialog for `doing tag -a -c 0` (autotag all)

### 1.0.47

- Remove check for file existence before attempting to run run_after script
- Don't autotag entries restarted with `again/resume`
- Add short flags (`-b`) for `--back` on all commands that support it

### 1.0.46

- Code cleanup

### 1.0.45

- Only execute run_after script if changes are written

### 1.0.44

- Remove unneccessary console logging

### 1.0.43

- Add `again` command to repeat last entry without @done tag
- Add `run_after` configuration option to execute external script after any change

### 1.0.42

- Fix note indentation in doing file

### 1.0.41

- Fix for repeated backreferences in tag transform

### 1.0.40

- Add `--tag_sort` to all subcommands with `--totals` option

### 1.0.39

- Tag transforms
- Option to sort tags by name in --totals

### 1.0.33

- Gem dependency updates

### 1.0.30

- Fix for array comparison error

### 1.0.29

- Bugfixes

### 1.0.28

- Global option `-x` to skip autotags and default_tags from global/local .doingrc
- Remove extra spaces when creating entry

### 1.0.27

- More graceful writing of default config (~/.doingrc) on first run
- Repaired testing setup. Needs moar tests.

### 1.0.26

- Add `--at` flag for `doing done`, e.g. `doing done --at=1:35pm --took=15m A new task I already finished`
- Allow decimal quantities when using natural language for hours or days, e.g. `--took=2.5h`
- Add `did` as a synonym for `done` subcommand

### 1.0.25

- Smarter method of getting user $HOME
- Improved avoiding duplicate tags when autotagging
- Improved autotag reporting

### 1.0.24

- `doing note` operates on whatever is most recent, not just the last note in Currently
- `doing tag` with no count specified operates on most recent entry in any section, not just Currently
- `doing tag` with a count greater than 1 requires a section to be specified
- Improved results reporting for `doing tag`

doing.completion.bash
# Bash completion for `doing` <http://brettterpstra.com/projects/doing>
# Completes commands, views, and section titles

_doing_show() {
	OLD_IFS="$IFS"
	local token=${COMP_WORDS[$COMP_CWORD]}
	IFS=$'\t'
	local words=$(doing sections)
	IFS="$OLD_IFS"


	if [[ "$token" == --* ]]; then
		COMPREPLY=( $( compgen -W '--bool --count --age --sort --times --totals --csv' -- $token ) )
	elif [[ "$token" == -* ]]; then
		COMPREPLY=( $( compgen -W '-b -c -a -s -t --bool --count --age --sort --times --totals --csv' -- $token ) )
	else

		local nocasematchWasOff=0
		shopt nocasematch >/dev/null || nocasematchWasOff=1
		(( nocasematchWasOff )) && shopt -s nocasematch

		local w matches=()
		OLD_IFS="$IFS"
		IFS=$'\t'‰

		for w in $words; do
			if [[ "$w" == "$token"* ]]; then
				matches+=("${w// /\ }")
			fi
		done

		IFS="$OLD_IFS"

		(( nocasematchWasOff )) && shopt -u nocasematch

		COMPREPLY=("${matches[@]}")
	fi

}

_doing_view() {
	OLD_IFS="$IFS"
	local token=${COMP_WORDS[$COMP_CWORD]}
	IFS=$'\t'
	local words=$(doing views)
	IFS="$OLD_IFS"

	if [[ "$token" == --* ]]; then
		COMPREPLY=( $( compgen -W '--section --count --csv --times --totals' -- $token ) )
	elif [[ "$token" == -* ]]; then
		COMPREPLY=( $( compgen -W '-s -c -t --section --count --csv --times --totals' -- $token ) )
	else
		local nocasematchWasOff=0
		shopt nocasematch >/dev/null || nocasematchWasOff=1
		(( nocasematchWasOff )) && shopt -s nocasematch

		OLD_IFS="$IFS"
		IFS=$'\t'
		local w matches=()
		for w in $words; do
			if [[ "$w" == "$token"* ]]; then matches+=("$w"); fi
		done
		IFS="$OLD_IFS"

		(( nocasematchWasOff )) && shopt -u nocasematch

		COMPREPLY=("${matches[@]}")
	fi

}

_doing()
{
	local last="${@: -1}"
	local token=${COMP_WORDS[$COMP_CWORD]}

	if [[ $last == "view" ]]; then
		_doing_view
	elif [[ $last =~ (show|archive|-s|--s|-t) ]]; then
		_doing_show
	else
		OLD_IFS="$IFS"
		IFS=$'\n'
		COMPREPLY=( $(compgen -W "$(doing help -c)" -- $token) )
		IFS="$OLD_IFS"
	fi
}

complete -F _doing doing
